std-bits/base/src/main/java/org/enso/base/enso_cloud/logging/LogMessage.java

```diff
@@ -0,0 +1,102 @@
+package org.enso.base.enso_cloud.logging;
+
+import com.fasterxml.jackson.databind.node.JsonNodeFactory;
+import com.fasterxml.jackson.databind.node.NullNode;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.fasterxml.jackson.databind.node.TextNode;
+import java.util.Objects;
+import org.enso.base.CurrentEnsoProject;
+import org.enso.base.enso_cloud.CloudAPI;
+
+/** Base class for log messages that are passed to the OpenSearch cloud endpoint. */
+public abstract class LogMessage {
+  /**
+   * A reserved field that is currently added by the cloud backend. Duplicating it will lead to
+   * internal server errors and log messages being discarded.
+   */
+  private static final String RESERVED_TYPE = "type";
+
+  private static final String PROJECT_NAME = "projectName";
+  private static final String PROJECT_ID = "projectId";
+  private static final String PROJECT_SESSION_ID = "projectSessionId";
+  private static final String LOCAL_TIMESTAMP = "localTimestamp";
+
+  private final String message;
+  private final String projectId;
+  private final String projectName;
+  private final String projectSessionId;
+
+  protected LogMessage(String message) {
+    this.message = Objects.requireNonNull(message);
+    this.projectId = CloudAPI.getCloudProjectId();
+    var currentProject = CurrentEnsoProject.get();
+    this.projectName = currentProject == null ? null : currentProject.fullName();
+    this.projectSessionId = CloudAPI.getCloudSessionId();
+  }
+
+  private static void checkNoRestrictedField(ObjectNode metadata, String fieldName) {
+    if (metadata.has(fieldName)) {
+      throw new IllegalArgumentException(
+          "Metadata cannot contain a field named '" + fieldName + "'. Metadata was: " + metadata);
+    }
+  }
+
+  protected static void checkNoRestrictedFields(ObjectNode metadata) {
```

Comment:
```suggestion
  static void checkNoRestrictedFields(ObjectNode metadata) {
```
I am not sure why this method is `protected`? In an API `protected static` of a subclassable class _protects nothing_. Everyone can make a subclass and call the method anytime.

Reply:
This suggestion causes compilation failure. This method is called on purpose from its subclass in https://github.com/enso-org/enso/blob/6ecdd77b5b24a00ce7839b8921f752d804dd0597/std-bits/base/src/main/java/org/enso/base/enso_cloud/audit/AuditLogMessage.java#L30

Reply:
I see. In such case the method should be `private` and be called from constructor of `LogMessage` to verify its `metadata` argument.